include/deal.II/lac/tensor_product_matrix.h

```diff
@@ -0,0 +1,371 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2017 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__tensor_product_matrix_h
+#define dealii__tensor_product_matrix_h
+
+
+#include <deal.II/base/config.h>
+#include <deal.II/lac/lapack_full_matrix.h>
+#include <deal.II/matrix_free/tensor_product_kernels.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+template <typename> class Vector;
+template <typename> class FullMatrix;
+
+/**
+ * This is a special matrix class defined as the tensor product (or Kronecker
+ * product) of 1D matrices of the type
+ * @f{align*}{
+ * L &= A \otimes M + M \otimes A
+ * @f}
+ * in 2D and
+ * @f{align*}{
+ * L &= A \otimes M \otimes + M \otimes A \otimes M + M \otimes M \otimes M
```

Comment:
This should be `A x M x M + M x A x M + M x M x A`, right? The last term has three `M`s.